### PR TITLE
Save MQTT persistent session to settings

### PR DESCRIPTION
**Pull request by jayteemo**
An MQTT session should only be considered valid after all subscriptions have completed.
Save a valid session flag to settings so it can be properly restored.
CIA-76

Update:  Based on discussion from PR #2849, use the zephyr mqtt clean session config instead of multiple persistent session configs.

### CODE REVIEW

**Comment by jayteemo** [comment]
Please advise on the best way to handle the Kconfig setup.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c [hunk 1]
```diff
@@ -26,6 +27,8 @@
 
 LOG_MODULE_REGISTER(nrf_cloud_transport, CONFIG_NRF_CLOUD_LOG_LEVEL);
 
+#define UNUSED_VARIABLE(X) ((void)(X))
```
Comment (jtguggedal) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:30]: The Zephyr macro `ARG_UNUSED(x)` can also be used for this purpose

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c [hunk 2]
```diff
@@ -608,6 +618,73 @@ static void aws_fota_cb_handler(struct aws_fota_event *fota_evt)
...
+static int nct_settings_set(const char *key, size_t len_rd,
+				 settings_read_cb read_cb, void *cb_arg)
+{
+	if (!key) {
+		return -EINVAL;
+	}
+
+	int read_val;
+
+	LOG_DBG("Settings key: %s", key);
```
Comment (jtguggedal) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:630]: missing `log_strdup()`

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c [hunk 3]
```diff
@@ -608,6 +618,73 @@ static void aws_fota_cb_handler(struct aws_fota_event *fota_evt)
...
+	ret = settings_save_one(SETTINGS_FULL_PERSISTENT_SESSION,
+				&session_valid, sizeof(session_valid));
+#endif
+	return ret;
+}
+
+static int nct_settings_init(void)
+{
+	int err = 0;
+
```
Comment (jtguggedal) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:661]: The `if defined` can be moved up here to not need the `UNUSED_VARIABLE()` macro

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c [hunk 4]
```diff
@@ -608,6 +618,73 @@ static void aws_fota_cb_handler(struct aws_fota_event *fota_evt)
 }
 #endif /* defined(CONFIG_AWS_FOTA) */
 
+static int nct_settings_set(const char *key, size_t len_rd,
```
Comment (jtguggedal) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:621]: I'm very confused by the settings API.. This is a function for reading the settings, right, but it's registered as a `.h_set` handler?

**Review by jtguggedal — COMMENTED** [review]
The settings API is a big mystery to me, but the PR seems to do the job when testing.
Did you test it with Thingy:91 + BSEC?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c (hunk 1 above)
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:30]: ah thanks, i thought there was a common one defined somewhere

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c (hunk 3 above)
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:661]: but then there would be an unused function warning for `nct_settings_set` 🙃

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c (hunk 4 above)
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:621]: yes i don't know why it is like that either 🤷‍♂️ 

[h_set](https://docs.zephyrproject.org/latest/reference/settings/index.html?highlight=settings_subsys_init#handlers)  - This gets called when the value is loaded from persisted storage with **settings_load()**, or when using settings_runtime_set() from the runtime backend.

**Comment by jayteemo** [comment]
> Did you test it with Thingy:91 + BSEC?

no, i'll try that now

**Comment by plskeggs** [comment]
Looks good.  How goes testing?

**Comment by jayteemo** [comment]
> How goes testing?

😑 
when running Thingy + BSEC, it crashes when BSEC calls `settings_register`
seems to be crashing at the `SYS_SLIST_FOR_EACH_CONTAINER` call in `settings_register`

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c (hunk 3 above)
Comment (jtguggedal) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:659]: Oh, that's true :(

**Comment by jayteemo** [comment]
crash fixed 🎈

**Review by jtguggedal — APPROVED** [review]
LGTM, thanks!

**Review by plskeggs — APPROVED** [review]
LGTM

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c [hunk 5]
```diff
@@ -608,6 +622,66 @@ static void aws_fota_cb_handler(struct aws_fota_event *fota_evt)
...
+}
+
+int save_session_state(const int session_valid)
+{
+	int ret = 0;
+
+#if defined(CONFIG_CLOUD_PERSISTENT_SESSIONS)
+	LOG_DBG("Setting session state: %d", session_valid);
+	persistent_session = (bool)session_valid;
+	ret = settings_save_one(SETTINGS_FULL_PERSISTENT_SESSION,
```
Comment (plskeggs) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:657]: I am wondering if it makes sense to not write the settings if they have not changed, in order to reduce unnecessary Flash wear?

Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:657]: correct, there is no explicit check in this function, however the function only gets called when the session state is changing.
also `settings_fcb_save` does such a check before writing.

**Comment by plskeggs** [comment]
@jayteemo, I just noticed another place that can skip subscriptions even with your changes.  I verified last week that actually AWS violates the MQTT spec and will return in the CONNACK message the session_present_flag = true even if the MQTT CONNECT message set clean_session = 1.

The place that still has a problem is aws_fota.c:aws_fota_mqtt_evt_handler().  It is passed the CONNACK message by nrf_cloud_transport.c:nct_mqtt_evt_handler():
`	err = aws_fota_mqtt_evt_handler(mqtt_client, _mqtt_evt);`

and uses the session_present_flag to skip subscribing to FOTA topics on line 401:
`		} else {
			LOG_DBG("Previous session valid;"
				" skipping FOTA subscriptions");
`

**Comment by jayteemo** [comment]
> It is passed the CONNACK message by nrf_cloud_transport.c:nct_mqtt_evt_handler()

so which config flag should aws_fota use to determine if persistent sessions should be used 😂

**Comment by plskeggs** [comment]
To simplify the decision making, I will forgo an opinion.  Too many cooks.

**Comment by jtguggedal** [comment]
CI fails for QEMU with this one:
`<err> Failed to save session state`

**Comment by jayteemo** [comment]
> Should there be some update to `prj_qemu_x86.conf` as well?

yes looks like it.  i will just enable clean sessions for now.
how does the settings module work for qemu?  what storage backend is required?

**Comment by jtguggedal** [comment]
> > Should there be some update to `prj_qemu_x86.conf` as well?
> 
> yes looks like it. i will just enable clean sessions for now.
> how does the settings module work for qemu? what storage backend is required?

This is what I've used for settings subsystem on QEMU, but I guess also variants would work:
```
CONFIG_FLASH=y
CONFIG_FLASH_MAP=y
CONFIG_FCB=y
CONFIG_SETTINGS=y
CONFIG_SETTINGS_FCB=y
```

File: applications/asset_tracker/prj_qemu_x86.conf [hunk 6]
```diff
@@ -25,6 +25,13 @@ CONFIG_NRF_CLOUD_CERTIFICATES_FILE="certificates.h"
 CONFIG_NRF_CLOUD_SEC_TAG=1
 CONFIG_CLOUD_API=y
 CONFIG_NRF_CLOUD_CONNECTION_POLL_THREAD=y
+CONFIG_MQTT_CLEAN_SESSION=n
+#s Settings for nRF Cloud persistent sessions (no clean session)
```
Comment (jtguggedal) [inline applications/asset_tracker/prj_qemu_x86.conf:29]: Stray `s` at the start of the comment

Comment (jtguggedal) [reply applications/asset_tracker/prj_qemu_x86.conf:29]: And also, a blank line before the settings session could be added

**Review by rlubos — APPROVED** [review]

**Pull request merged**